074 dirt texture erasing

CrispyKorn opened the pull request:
Feature Components:
- Create tool scriptable objects and use any sprite for it's erasing mask.
- Erase textures using any tool.
- Adjust the size and strength of each tool from within the scriptable object.
- Duplicate any erasable texture as much as you like.

Note: Objects who's sprite you want to be erasable must be tagged 'Erasable'.

📝 Checklist before requesting a review:
- [x]  I have run this branch locally
- [x]  I have performed a self-review of my code
- [x]  I have confirmed critical features still function

SkyRiser568 commented on MonstieWash/Assets/Scripts/Eraser.cs:
```diff
@@ -0,0 +1,233 @@
+    private Sprite DuplicateSprite(Sprite sprite)
+    {
+        var textureCopy = new Texture2D(sprite.texture.width, sprite.texture.height, sprite.texture.format, false);
+        textureCopy.alphaIsTransparency = true;
```
Texture2D.alphaIsTransparency is editor only and can't be changed unless done in TextureImporter.alphaIsTransparency. Also it breaks exporting builds

CrispyKorn replied:
Hmm that's v unfortunate. I'll have a look at another way to do this that is legal.

SkyRiser568 commented on MonstieWash/Assets/Scripts/Eraser.cs:
```diff
@@ -0,0 +1,233 @@
+    /// <returns>Whether or not the mouse moved since last frame.</returns>
+    private bool MouseMoved()
+    {
+        var transformScreenPoint = Camera.main.WorldToScreenPoint(transform.position);
+        m_mouseScreenPos = new Vector2Int(Mathf.RoundToInt(transformScreenPoint.x), Mathf.RoundToInt(transformScreenPoint.y));
```
Would be good to have this follow the hand instead of the mouse so we don't have 2 different controls

CrispyKorn replied:
Ah yep most of this was done before your feature was added just for testing, I can update it to include your work. 👍

SkyRiser568 commented on MonstieWash/Assets/Scripts/FollowMouse.cs:
```diff
@@ -0,0 +1,26 @@
+
+    private void Update()
+    {
+        Vector3 mousePos = Camera.main.ScreenToWorldPoint(Input.mousePosition);
```
same as other comment to follow hand instead of mouse

SkyRiser568 commented on MonstieWash/Assets/Scripts/Tool.cs:
```diff
@@ -0,0 +1,30 @@
+
+    public void Initialize()
+    {
+        strength = Mathf.Pow(1.0472f, inputStrength);
```
1.0472 should probably be explained,  eg. scaling factor, also does inputStrength need to be adjusted

CrispyKorn replied:
Ah good catch, will do. I'll explain more of what each variable does in the code (the 1.0472 is just 100 to the 100th root), but for now the inputStrength is what the user inputs on a slider (0 - 100) which represents the strength/opacity of the tool and then that is scaled to make the strength adjustments seem more linear and intuitive. Then strength is the final (actual) tool strength value. These are mainly settings for designers to use when setting up tools.

SkyRiser568 commented on MonstieWash/Assets/Scripts/Tool.cs:
```diff
@@ -0,0 +1,30 @@
+[CreateAssetMenu(fileName = "Tool", menuName = "ScriptableObjects/Tool")]
+public class Tool : ScriptableObject
+{
+    public Sprite mask;
+    [HideInInspector]
+    public float strength;
+    [Range(1, 10)] 
+    public int size = 1;
+    [HideInInspector] 
+    public byte[] maskPixels;
```
Could separate what is/isn't shown in inspector

CrispyKorn replied:
Will do, good idea! I thought that was messy but wasn't sure the best way to lay it out.

SkyRiser568 commented on MonstieWash/Assets/Scenes/MonstieWash.unity:
```diff
@@ -297,9 +325,270 @@ Transform:
+  - component: {fileID: 2064844344}
+  m_Layer: 0
+  m_Name: Dirt
+  m_TagString: Untagged
```
Dirt object is untagged
should be Erasable?
```suggestion
  m_TagString: Erasable
```

CrispyKorn replied:
Oh this was just for demonstration of how the tag works, these objects aren't important (just for testing) and will most likely be deleted anyways.

SkyRiser568 commented on MonstieWash/Assets/Scripts/Eraser.cs:
```diff
@@ -0,0 +1,233 @@
+        }
+    }
+
+    /// <summary>
+    /// Sets up the tool and related variables ready for use.
+    /// </summary>
```
If we'll need to document our functions with XML tags it would be nice to have it as part of the conventions in the dev wiki

CrispyKorn replied:
Ah yep, good idea. I think I'll add it there as it's v helpful when reading through the code.

SkyRiser568 reviewed (changes requested):
Great stuff! like WOW. 
Line 99 has something that causes error when trying to export a build.
Then there's some other changes/cleanup for you to go over.
😄

CrispyKorn commented:
I've fixed up the code based on your feedback and made my system work with yours. Also, turns out I didn't even need the line that broke the build after all .-. so I just removed it. Take another look and again let me know if you'd like anything changed!

SkyRiser568 commented on MonstieWash/Assets/Scripts/PlayerHand.cs:
```diff
@@ -4,49 +4,42 @@
+        var cameraHeightInWorldUnits = Camera.main.orthographicSize;
+
+        //Keep within screen bounds
+        newPosition.x = Mathf.Clamp(newPosition.x, -cameraWidthInWorldUnits, cameraWidthInWorldUnits);
+        newPosition.y = Mathf.Clamp(newPosition.y, -cameraHeightInWorldUnits, cameraHeightInWorldUnits);
```
thanks for this, way cleaner method that I forgot about

SkyRiser568 commented on MonstieWash/Assets/Scripts/PlayerHand.cs:
```diff
@@ -4,49 +4,42 @@
+    private Vector3 CalculateMovement()
+    {
+        var newPosition = transform.position + new Vector3(m_moveHorizontal * cursorSpeed * Time.deltaTime, m_moveVertical * cursorSpeed * Time.deltaTime, 0f);
+        var aspectRatio = 16f / 9f;
```
I think you could use Camera.aspect here for if the aspect ratio is changed

CrispyKorn replied:
Good catch!

SkyRiser568 commented on MonstieWash/Assets/Scripts/PlayerHand.cs:
```diff
@@ -4,49 +4,42 @@
-        
-        m_transform.position = transform.position + new Vector3(m_moveHorizontal * cursorSpeed * Time.deltaTime, m_moveVertical * cursorSpeed * Time.deltaTime, 0);
+        transform.position = CalculateMovement();
+        handPosition = Camera.main.WorldToScreenPoint(transform.position);
```
This one is my bad, handPosition should probably be assigned in world point because it will need to be compared to many other objects in scene which also default to world point. I only changed it to screen point because I had to compare to screen edges

CrispyKorn replied:
Ah I see, well in that case hand position can be found just from transform.position, so would we want to still keep handPosition at all? I use it currently in the eraser code but I don't necessarily need it.

SkyRiser568 replied:
i'd remove it since its not needed

SkyRiser568 commented on MonstieWash/Assets/Scripts/FollowMouse.cs:
I don't think this script is being used anymore or I just didn't see where. If it's deprecated it would be good to leave a note in the script for why.

CrispyKorn replied:
Yep it's from before your code was done, I've just deleted it since it's not used for anything anymore.

SkyRiser568 reviewed (changes requested):
Changes are good, there are a few bits of cleanup I think would be good before merging.

SkyRiser568 commented on MonstieWash/Assets/Scripts/Eraser.cs:
```diff
@@ -0,0 +1,244 @@
+using System;
+using System.Collections.Generic;
+using UnityEditor;
```
this should just be about line 3 with using UnityEditor

SkyRiser568 commented:
Since these changes should be simple, if there's nothing new that needs review after you're done. I reckon you can just merge it to dev

CrispyKorn commented:
Actually I just added the Input System so if you could take a look at that just to make sure it's all okay (it all works great in testing) that would be great! After that we should be good to merge.

CrispyKorn commented:
Also if possible it'd be great if you could create a new "approve" review when you're happy with it just for the sake of the process (which I will update on notion when I get the chance) and so the status of the pr is updated. Thanks!!

SkyRiser568 reviewed (approved):
The requested changes are all good. Merge can be done after looking at the comments.

I think changing the input system now was a bit out of scope for this feature and premature, I would have liked some input from other programmers before changing systems and maybe even from the designers because it will impact them down the line. 
I don't think its worth reverting anything that has been done but we should schedule a meeting for it.

CrispyKorn commented:
> The requested changes are all good. Merge can be done after looking at the comments.
> 
> I think changing the input system now was a bit out of scope for this feature and premature, I would have liked some input from other programmers before changing systems and maybe even from the designers because it will impact them down the line. I don't think its worth reverting anything that has been done but we should schedule a meeting for it.

I understand, I thought it was a given we would use the Input System as that's the proper way to do controls nowadays but you're right that it was out of scope for this feature, I just wanted to get it out of the way early. I'll change how its implemented and merge the changes. Thanks for your patience!
